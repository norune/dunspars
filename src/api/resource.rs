use std::collections::HashMap;

use anyhow::{anyhow, bail, Result};
use futures::stream::FuturesOrdered;
use futures::StreamExt;
use regex::Regex;

use rustemon::client::RustemonClient;
use rustemon::games::version_group as rustemon_version;

use super::utils;
use crate::pokemon::Game;

pub enum ResourceResult {
    Valid,
    Invalid(Vec<String>),
}

#[allow(async_fn_in_trait)]
pub trait Resource: Sized {
    fn get_matches(&self, value: &str) -> Vec<String> {
        self.resource()
            .iter()
            .filter_map(|r| {
                let close_enough = if !r.is_empty() && !value.is_empty() {
                    let first_r = r.chars().next().unwrap();
                    let first_value = value.chars().next().unwrap();

                    // Only perform spellcheck on first character match; potentially expensive
                    first_r == first_value && strsim::levenshtein(r, value) < 4
                } else {
                    false
                };

                if r.contains(value) || close_enough {
                    Some(r.clone())
                } else {
                    None
                }
            })
            .collect::<Vec<String>>()
    }

    fn check(&self, value: &str) -> ResourceResult {
        let matches = self.get_matches(value);
        if matches.iter().any(|m| *m == value) {
            ResourceResult::Valid
        } else {
            ResourceResult::Invalid(matches)
        }
    }

    fn validate(&self, value: &str) -> Result<String> {
        let value = value.to_lowercase();
        match self.check(&value) {
            ResourceResult::Valid => Ok(value),
            ResourceResult::Invalid(matches) => bail!(Self::invalid_message(&value, &matches)),
        }
    }

    fn invalid_message(value: &str, matches: &[String]) -> String {
        let resource_name = Self::label();
        let mut message = format!("{resource_name} '{value}' not found.");

        if matches.len() > 20 {
            message += " Potential matches found; too many to display.";
        } else if !matches.is_empty() {
            message += &format!(" Potential matches: {}.", matches.join(" "));
        }

        message
    }

    async fn try_new(api: &RustemonClient) -> Result<Self>;
    fn resource(&self) -> Vec<String>;
    fn label() -> &'static str;
}

#[derive(Debug)]
pub struct PokemonResource {
    resource: Vec<String>,
}
impl Resource for PokemonResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_pokemon(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Pokémon"
    }
}

#[derive(Debug)]
pub struct TypeResource {
    resource: Vec<String>,
}
impl Resource for TypeResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_types(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Type"
    }
}

#[derive(Debug)]
pub struct MoveResource {
    resource: Vec<String>,
}
impl Resource for MoveResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_moves(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Move"
    }
}

#[derive(Debug)]
pub struct AbilityResource {
    resource: Vec<String>,
}
impl Resource for AbilityResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_abilities(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Ability"
    }
}

#[derive(Debug)]
pub struct GameResource {
    resource: HashMap<String, Game>,
    gen_url_regex: Regex,
}
impl Resource for GameResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        // PokéAPI keeps generation names in Roman numerals.
        // Might be quicker to just take it from resource urls via regex instead.
        // Regex compilation is expensive, so we're compiling it just once here.
        let gen_url_regex = Regex::new(r"generation/(?P<gen>\d+)/?$").unwrap();

        let mut resource = HashMap::new();
        let game_names = utils::get_all_games(client).await?;
        let game_data_futures: FuturesOrdered<_> = game_names
            .iter()
            .map(|g| rustemon_version::get_by_name(g, client))
            .collect();
        let game_data: Vec<_> = game_data_futures.collect().await;

        for (i, game) in game_data.into_iter().enumerate() {
            let game = game?;
            let generation = capture_gen_url(&game.generation.url, &gen_url_regex).unwrap();
            resource.insert(game.name.clone(), Game::new(game.name, i as u8, generation));
        }

        Ok(Self {
            resource,
            gen_url_regex,
        })
    }

    fn resource(&self) -> Vec<String> {
        let mut games = self.resource.iter().map(|r| r.1).collect::<Vec<&Game>>();
        games.sort_by_key(|g| g.order);

        games
            .iter()
            .map(|g| g.name.clone())
            .collect::<Vec<String>>()
    }

    fn label() -> &'static str {
        "Game"
    }
}
impl GameResource {
    pub fn get_gen(&self, game: &str) -> u8 {
        self.resource.get(game).unwrap().generation
    }

    pub fn get_gen_from_url(&self, url: &str) -> u8 {
        capture_gen_url(url, &self.gen_url_regex).unwrap()
    }
}

fn capture_gen_url(url: &str, gen_url_regex: &Regex) -> Result<u8> {
    if let Some(caps) = gen_url_regex.captures(url) {
        Ok(caps["gen"].parse::<u8>()?)
    } else {
        Err(anyhow!("Generation not found in resource url"))
    }
}
