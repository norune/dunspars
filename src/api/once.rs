use crate::resource::GameResource;
use regex::Regex;
use std::sync::OnceLock;

pub fn game_resource() -> &'static GameResource {
    static GAME_RESOURCE: OnceLock<GameResource> = OnceLock::new();
    GAME_RESOURCE.get_or_init(|| GameResource::try_new().unwrap())
}

pub fn gen_url_regex() -> &'static Regex {
    static GEN_URL_REGEX: OnceLock<Regex> = OnceLock::new();
    GEN_URL_REGEX.get_or_init(|| {
        // Pok√©API keeps generation names in Roman numerals.
        // Might be quicker to just take it from resource urls via regex instead.
        // Regex compilation is expensive, so we're compiling it just once here.
        Regex::new(r"generation/(?P<gen>\d+)/?$").unwrap()
    })
}
