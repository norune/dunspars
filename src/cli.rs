mod display;
mod utils;

use anyhow::{anyhow, Result};
use clap::{Parser, Subcommand};
use indoc::{formatdoc, printdoc};

use crate::data::api::resource::{
    AbilityResource, GameResource, GetGeneration, MoveResource, PokemonResource, Resource,
    TypeResource,
};
use crate::data::api::ApiWrapper;
use crate::data::{Ability, Move, Pokemon, PokemonData, Type};
use display::*;

const VERSION: &str = env!("DUNSPARS_VERSION");

#[derive(Parser)]
#[command(author, version = VERSION, about, long_about = None)]
struct Cli {
    #[command(subcommand)]
    command: Commands,
    /// Sets the mainline Pokémon game the output will be based on
    #[clap(short, long, global = true)]
    game: Option<String>,
    /// Force output to include colors
    #[clap(long, action = clap::ArgAction::SetTrue, global = true)]
    color: bool,
    /// Force output to exclude colors
    #[clap(long, action = clap::ArgAction::SetTrue, global = true)]
    no_color: bool,
}

#[derive(Subcommand)]
enum Commands {
    /// Prints general data about a Pokémon
    Pokemon {
        /// Name of the Pokémon
        pokemon: String,
        /// Display all move data the Pokémon is capable of learning
        #[arg(short, long, action = clap::ArgAction::SetTrue)]
        moves: bool,
        /// Display the Pokémon evolutionary line
        #[arg(short, long, action = clap::ArgAction::SetTrue)]
        evolution: bool,
    },
    /// Prints matchup data between Pokémon
    Match {
        /// Names of the defending Pokémon; max 6
        #[arg(required = true, num_args = 1..=6)]
        defenders: Vec<String>,
        /// Name of the attacking Pokémon
        attacker: String,
        /// Display only moves that match the user's type
        #[arg(short, long, action = clap::ArgAction::SetTrue)]
        stab_only: bool,
        /// Display verbose output
        #[arg(short, long, action = clap::ArgAction::SetTrue)]
        verbose: bool,
    },
    /// Prints type coverage based on the provided Pokémon
    Coverage {
        /// Names of Pokémon; max 6
        #[arg(required = true, num_args = 1..=6)]
        pokemon: Vec<String>,
    },
    /// Prints data about a Pokémon type
    Type {
        /// Name of the type
        type_: String,
    },
    /// Prints data about a Pokémon move
    Move {
        /// Name of the move
        move_: String,
    },
    /// Prints data about a Pokémon ability
    Ability {
        /// Name of the ability
        ability: String,
    },
    /// Prints all possible names from a Resource such as Pokémon, Moves, etc
    Resource {
        /// Name of the resource
        #[arg(value_enum)]
        resource: ResourceArgs,
        /// Value to be printed in between values. Defaults to newline
        #[arg(short, long)]
        delimiter: Option<String>,
    },
    /// Actions regarding the program's cache
    Cache {
        /// Action to undertake
        #[arg(value_enum)]
        action: CacheAction,
    },
}

#[derive(Clone, clap::ValueEnum)]
enum ResourceArgs {
    Pokemon,
    Moves,
    Abilities,
    Games,
    Types,
}

#[derive(Clone, clap::ValueEnum)]
enum CacheAction {
    Clear,
}

pub async fn run() -> Result<()> {
    let cli = Cli::parse();
    let api = ApiWrapper::try_new().await?;

    let mut config_builder = ConfigBuilder::new(&api.game_resource);
    if let Some(game) = cli.game {
        config_builder = config_builder.game(game);
    }
    if cli.color {
        config_builder = config_builder.color_enabled(true);
    }
    if cli.no_color {
        config_builder = config_builder.color_enabled(false);
    }
    let config = config_builder.build()?;

    let program = Program::new(config, api);

    match cli.command {
        Commands::Pokemon {
            pokemon,
            moves,
            evolution,
        } => print!("{}", program.run_pokemon(pokemon, moves, evolution).await?),
        Commands::Type { type_ } => print!("{}", program.run_type(type_).await?),
        Commands::Move { move_ } => print!("{}", program.run_move(move_).await?),
        Commands::Ability { ability } => print!("{}", program.run_ability(ability).await?),
        Commands::Match {
            defenders,
            attacker,
            stab_only,
            verbose,
        } => print!(
            "{}",
            program
                .run_match(defenders, attacker, verbose, stab_only)
                .await?
        ),
        Commands::Coverage { pokemon } => print!("{}", program.run_coverage(pokemon).await?),
        Commands::Resource {
            resource,
            delimiter,
        } => program.run_resource(resource, delimiter).await?,
        Commands::Cache { action } => program.run_cache(action).await?,
    }

    Ok(())
}

struct ConfigBuilder<'a> {
    game: Option<String>,
    color_enabled: Option<bool>,
    game_resource: &'a GameResource,
}

impl<'a> ConfigBuilder<'a> {
    pub fn new(game_resource: &'a GameResource) -> Self {
        ConfigBuilder {
            game_resource,
            game: None,
            color_enabled: None,
        }
    }

    pub fn game(mut self, game: String) -> Self {
        self.game = Some(game);
        self
    }

    pub fn color_enabled(mut self, color_enabled: bool) -> Self {
        self.color_enabled = Some(color_enabled);
        self
    }

    pub fn build(self) -> Result<Config> {
        let game = match self.game {
            Some(game) => self.game_resource.validate(&game)?,
            None => self
                .get_latest_game()
                .ok_or(anyhow!("Cannot find the latest game"))?,
        };

        let generation = self.game_resource.get_gen(&game);
        let color_enabled = self.color_enabled.unwrap_or(utils::is_color_enabled());

        Ok(Config {
            game,
            generation,
            color_enabled,
        })
    }

    fn get_latest_game(&self) -> Option<String> {
        self.game_resource.resource().last().map(|g| g.to_string())
    }
}

struct Config {
    game: String,
    generation: u8,
    color_enabled: bool,
}

struct Program {
    config: Config,
    api: ApiWrapper,
}

impl Program {
    pub fn new(config: Config, api: ApiWrapper) -> Self {
        Self { config, api }
    }

    async fn run_pokemon(&self, name: String, moves: bool, evolution: bool) -> Result<String> {
        let resource = PokemonResource::try_new(&self.api.client).await?;
        let pokemon_name = resource.validate(&name)?;

        let pokemon = PokemonData::from_name(&self.api, &pokemon_name, &self.config.game).await?;
        let pokemon_display = DisplayComponent::new(&pokemon, self.config.color_enabled);

        let defense_chart = pokemon.get_defense_chart().await?;
        let defense_chart_ctx = TypeChartComponent {
            type_chart: &defense_chart,
        };
        let type_chart_display =
            DisplayComponent::new(defense_chart_ctx, self.config.color_enabled);

        let mut output = formatdoc! {
            "
            {pokemon_display}

            {type_chart_display}
            "
        };

        if evolution {
            let evolution_step = pokemon.get_evolution_steps().await?;
            let evolution_step_display =
                DisplayComponent::new(&evolution_step, self.config.color_enabled);
            output += formatdoc! {
                "

                {evolution_step_display}
                "
            }
            .as_str();
        }

        if moves {
            let moves = pokemon.get_moves().await?;
            let move_list_context = MoveListComponent {
                move_list: &moves,
                pokemon: &pokemon,
            };
            let move_list_display =
                DisplayComponent::new(move_list_context, self.config.color_enabled);

            output += formatdoc! {
                "

                {move_list_display}
                "
            }
            .as_str();
        }

        Ok(output)
    }

    async fn run_type(&self, name: String) -> Result<String> {
        let resource = TypeResource::try_new(&self.api.client).await?;
        let type_name = resource.validate(&name)?;

        let Type {
            offense_chart,
            defense_chart,
            ..
        } = Type::from_name(&self.api, &type_name, self.config.generation).await?;

        let offense_chart_ctx = TypeChartComponent {
            type_chart: &offense_chart,
        };
        let offense_chart_display =
            DisplayComponent::new(offense_chart_ctx, self.config.color_enabled);

        let defense_chart_ctx = TypeChartComponent {
            type_chart: &defense_chart,
        };
        let defense_chart_display =
            DisplayComponent::new(defense_chart_ctx, self.config.color_enabled);

        let output = formatdoc! {
            "
            {offense_chart_display}

            {defense_chart_display}
            "
        };

        Ok(output)
    }

    async fn run_match(
        &self,
        defender_names: Vec<String>,
        attacker_name: String,
        verbose: bool,
        stab_only: bool,
    ) -> Result<String> {
        let resource = PokemonResource::try_new(&self.api.client).await?;

        let attacker_name = resource.validate(&attacker_name)?;
        let attacker_data =
            PokemonData::from_name(&self.api, &attacker_name, &self.config.game).await?;
        let attacker_moves = attacker_data.get_moves().await?;
        let attacker_chart = attacker_data.get_defense_chart().await?;
        let attacker = Pokemon::new(attacker_data, attacker_chart, attacker_moves);

        let mut defenders = vec![];

        for defender_name in defender_names {
            let defender_name = resource.validate(&defender_name)?;
            let defender_data =
                PokemonData::from_name(&self.api, &defender_name, &self.config.game).await?;
            let defender_moves = defender_data.get_moves().await?;
            let defender_chart = defender_data.get_defense_chart().await?;
            let defender = Pokemon::new(defender_data, defender_chart, defender_moves);

            defenders.push(defender);
        }

        let mut output = String::from("");
        for defender in defenders {
            let match_context = MatchComponent {
                defender: &defender,
                attacker: &attacker,
                verbose,
                stab_only,
            };
            let match_display = DisplayComponent::new(match_context, self.config.color_enabled);

            output += formatdoc! {
                "
                {match_display}


                "
            }
            .as_str();
        }

        Ok(output)
    }

    async fn run_coverage(&self, names: Vec<String>) -> Result<String> {
        let resource = PokemonResource::try_new(&self.api.client).await?;
        let mut pokemon = vec![];

        for name in names {
            let name = resource.validate(&name)?;
            let data = PokemonData::from_name(&self.api, &name, &self.config.game).await?;
            let moves = data.get_moves().await?;
            let chart = data.get_defense_chart().await?;

            let mon = Pokemon::new(data, chart, moves);
            pokemon.push(mon);
        }

        let coverage_ctx = CoverageComponent::try_new(&pokemon).await?;
        let coverage_display = DisplayComponent::new(coverage_ctx, self.config.color_enabled);

        Ok(formatdoc! {
            "
            {coverage_display}
            "
        })
    }

    async fn run_move(&self, name: String) -> Result<String> {
        let resource = MoveResource::try_new(&self.api.client).await?;
        let move_name = resource.validate(&name)?;

        let move_ = Move::from_name(&self.api, &move_name, self.config.generation).await?;
        let move_display = DisplayComponent::new(&move_, self.config.color_enabled);

        let output = formatdoc! {
            "
            {move_display}
            "
        };

        Ok(output)
    }

    async fn run_ability(&self, name: String) -> Result<String> {
        let resource = AbilityResource::try_new(&self.api.client).await?;
        let ability_name = resource.validate(&name)?;

        let ability = Ability::from_name(&self.api, &ability_name, self.config.generation).await?;
        let ability_display = DisplayComponent::new(&ability, self.config.color_enabled);

        let output = formatdoc! {
            "
            {ability_display}
            "
        };

        Ok(output)
    }

    async fn run_resource(&self, resource: ResourceArgs, delimiter: Option<String>) -> Result<()> {
        let delimiter = delimiter.unwrap_or("\n".to_string());
        let resource = match resource {
            ResourceArgs::Pokemon => PokemonResource::try_new(&self.api.client)
                .await?
                .resource()
                .join(&delimiter),
            ResourceArgs::Moves => MoveResource::try_new(&self.api.client)
                .await?
                .resource()
                .join(&delimiter),
            ResourceArgs::Abilities => AbilityResource::try_new(&self.api.client)
                .await?
                .resource()
                .join(&delimiter),
            ResourceArgs::Types => TypeResource::try_new(&self.api.client)
                .await?
                .resource()
                .join(&delimiter),
            ResourceArgs::Games => GameResource::try_new(&self.api.client)
                .await?
                .resource()
                .join(&delimiter),
        };

        printdoc! {
            "
            {resource}
            "
        };

        Ok(())
    }

    async fn run_cache(&self, action: CacheAction) -> Result<()> {
        match action {
            CacheAction::Clear => self.api.clear_cache().await,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    async fn setup_program(game: &str) -> Program {
        let api = ApiWrapper::try_new().await.unwrap();
        let config = ConfigBuilder::new(&api.game_resource)
            .game(String::from(game))
            .color_enabled(false)
            .build()
            .unwrap();
        Program::new(config, api)
    }

    #[tokio::test]
    async fn run_pokemon() {
        let program = setup_program("scarlet-violet").await;

        let output = program
            .run_pokemon(String::from("ceruledge"), false, false)
            .await
            .unwrap();

        insta::with_settings!({
            description => "pokemon ceruledge --game scarlet-violet",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_pokemon_evolution() {
        let program = setup_program("sword-shield").await;

        let cascoon = program
            .run_pokemon(String::from("cascoon"), false, true)
            .await
            .unwrap();

        insta::with_settings!({
            description => "pokemon cascoon --evolution --game sword-shield",
            omit_expression => true
        }, {
            insta::assert_snapshot!(cascoon);
        });

        let politoed = program
            .run_pokemon(String::from("politoed"), false, true)
            .await
            .unwrap();

        insta::with_settings!({
            description => "pokemon politoed --evolution --game sword-shield",
            omit_expression => true
        }, {
            insta::assert_snapshot!(politoed);
        });

        let applin = program
            .run_pokemon(String::from("applin"), false, true)
            .await
            .unwrap();

        insta::with_settings!({
            description => "pokemon applin --evolution --game sword-shield",
            omit_expression => true
        }, {
            insta::assert_snapshot!(applin);
        });
    }

    #[tokio::test]
    async fn run_pokemon_moves() {
        let program = setup_program("scarlet-violet").await;

        let output = program
            .run_pokemon(String::from("blaziken"), true, false)
            .await
            .unwrap();

        insta::with_settings!({
            description => "pokemon blaziken --moves --game scarlet-violet",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_type() {
        let program = setup_program("platinum").await;

        let output = program.run_type(String::from("ice")).await.unwrap();

        insta::with_settings!({
            description => "type ice --game platinum",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_move() {
        let program = setup_program("sun-moon").await;

        let output = program.run_move(String::from("brick-break")).await.unwrap();

        insta::with_settings!({
            description => "move brick-break --game sun-moon",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_ability() {
        let program = setup_program("black-white").await;

        let output = program
            .run_ability(String::from("intimidate"))
            .await
            .unwrap();

        insta::with_settings!({
            description => "ability intimidate --game black-white",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_match() {
        let program = setup_program("x-y").await;
        let defenders = vec![String::from("golem"), String::from("pachirisu")];
        let attacker = String::from("lapras");

        let non_verbose = program
            .run_match(defenders.clone(), attacker.clone(), false, false)
            .await
            .unwrap();

        let stab_only = program
            .run_match(defenders.clone(), attacker.clone(), false, true)
            .await
            .unwrap();

        let verbose = program
            .run_match(defenders.clone(), attacker.clone(), true, false)
            .await
            .unwrap();

        insta::with_settings!({
            description => "match golem pachirisu lapras --game x-y",
            omit_expression => true
        }, {
            insta::assert_snapshot!(non_verbose);
        });

        insta::with_settings!({
            description => "match golem pachirisu lapras --stab-only --game x-y",
            omit_expression => true
        }, {
            insta::assert_snapshot!(stab_only);
        });

        insta::with_settings!({
            description => "match golem pachirisu lapras --verbose --game x-y",
            omit_expression => true
        }, {
            insta::assert_snapshot!(verbose);
        });
    }

    #[tokio::test]
    async fn run_coverage() {
        let program = setup_program("the-indigo-disk").await;
        let team = vec![
            String::from("flamigo"),
            String::from("cramorant"),
            String::from("ribombee"),
            String::from("ogerpon-cornerstone-mask"),
            String::from("dudunsparce"),
            String::from("sinistcha"),
        ];

        let output = program.run_coverage(team).await.unwrap();

        insta::with_settings!({
            description => "coverage flamigo cramorant ribombee ogerpon-cornerstone-mask dudunsparce sinistcha --game the-indigo-disk",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }
}
