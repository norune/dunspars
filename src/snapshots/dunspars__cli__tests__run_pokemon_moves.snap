---
source: src/cli.rs
description: pokemon blaziken --moves --game scarlet-violet
---
blaziken fire fighting 
blaze speed-boost(h)
hp    atk   def   satk  sdef  spd   total
80    120   70    110   70    80    530   
scarlet-violet gen-9

fire fighting defense
double: flying ground psychic water
neutral: dragon electric fairy fighting ghost normal poison rock
half: dark fire grass ice steel
quarter: bug

moves
counter(s)           fighting physical   power: N/A  accuracy: 100  pp: 20    egg 
crush-claw           normal physical     power: 75   accuracy: 95   pp: 10    egg 
ember(s)             fire special        power: 40   accuracy: 100  pp: 25    egg 
feather-dance        flying status       power: N/A  accuracy: 100  pp: 15    egg 
feint                normal physical     power: 30   accuracy: 100  pp: 10    egg 
fire-punch(s)        fire physical       power: 75   accuracy: 100  pp: 15    egg 
flamethrower(s)      fire special        power: 90   accuracy: 100  pp: 15    egg 
last-resort          normal physical     power: 140  accuracy: 100  pp: 5     egg 
night-slash          dark physical       power: 70   accuracy: 100  pp: 15    egg 
peck                 flying physical     power: 35   accuracy: 100  pp: 35    egg 
blaze-kick(s)        fire physical       power: 85   accuracy: 90   pp: 10    level-up evolve
double-kick(s)       fighting physical   power: 30   accuracy: 100  pp: 30    level-up 1
growl                normal status       power: N/A  accuracy: 100  pp: 40    level-up 1
quick-attack         normal physical     power: 40   accuracy: 100  pp: 30    level-up 1
scratch              normal physical     power: 40   accuracy: 100  pp: 35    level-up 1
flame-charge(s)      fire physical       power: 50   accuracy: 100  pp: 20    level-up 9
detect(s)            fighting status     power: N/A  accuracy: N/A  pp: 5     level-up 12
sand-attack          ground status       power: N/A  accuracy: 100  pp: 15    level-up 15
aerial-ace           flying physical     power: 60   accuracy: N/A  pp: 20    level-up 20
slash                normal physical     power: 70   accuracy: 100  pp: 20    level-up 25
bounce               flying physical     power: 85   accuracy: 85   pp: 5     level-up 30
focus-energy         normal status       power: N/A  accuracy: N/A  pp: 30    level-up 35
bulk-up(s)           fighting status     power: N/A  accuracy: N/A  pp: 20    level-up 42
reversal(s)          fighting physical   power: N/A  accuracy: 100  pp: 15    level-up 49
flare-blitz(s)       fire physical       power: 120  accuracy: 100  pp: 15    level-up 56
brave-bird           flying physical     power: 120  accuracy: 100  pp: 15    level-up 63
acrobatics           flying physical     power: 55   accuracy: 100  pp: 15    machine 
agility              psychic status      power: N/A  accuracy: N/A  pp: 30    machine 
aura-sphere(s)       fighting special    power: 80   accuracy: N/A  pp: 20    machine 
baton-pass           normal status       power: N/A  accuracy: N/A  pp: 40    machine 
blast-burn(s)        fire special        power: 150  accuracy: 90   pp: 5     machine 
body-slam            normal physical     power: 85   accuracy: 100  pp: 15    machine 
brick-break(s)       fighting physical   power: 75   accuracy: 100  pp: 15    machine 
bulldoze             ground physical     power: 60   accuracy: 100  pp: 20    machine 
close-combat(s)      fighting physical   power: 120  accuracy: 100  pp: 5     machine 
coaching(s)          fighting status     power: N/A  accuracy: N/A  pp: 10    machine 
curse                ghost status        power: N/A  accuracy: N/A  pp: 10    machine 
dig                  ground physical     power: 80   accuracy: 100  pp: 10    machine 
double-edge          normal physical     power: 120  accuracy: 100  pp: 15    machine 
earthquake           ground physical     power: 100  accuracy: 100  pp: 10    machine 
endure               normal status       power: N/A  accuracy: N/A  pp: 10    machine 
facade               normal physical     power: 70   accuracy: 100  pp: 20    machine 
fire-blast(s)        fire special        power: 110  accuracy: 85   pp: 5     machine 
fire-pledge(s)       fire special        power: 80   accuracy: 100  pp: 10    machine 
fire-spin(s)         fire special        power: 35   accuracy: 85   pp: 15    machine 
fling                dark physical       power: N/A  accuracy: 100  pp: 10    machine 
focus-blast(s)       fighting special    power: 120  accuracy: 70   pp: 5     machine 
focus-punch(s)       fighting physical   power: 150  accuracy: 100  pp: 20    machine 
giga-impact          normal physical     power: 150  accuracy: 90   pp: 5     machine 
heat-crash(s)        fire physical       power: N/A  accuracy: 100  pp: 10    machine 
heat-wave(s)         fire special        power: 95   accuracy: 90   pp: 10    machine 
helping-hand         normal status       power: N/A  accuracy: N/A  pp: 20    machine 
hyper-beam           normal special      power: 150  accuracy: 90   pp: 5     machine 
knock-off            dark physical       power: 65   accuracy: 100  pp: 20    machine 
low-kick(s)          fighting physical   power: N/A  accuracy: 100  pp: 20    machine 
low-sweep(s)         fighting physical   power: 65   accuracy: 100  pp: 20    machine 
mud-slap             ground special      power: 20   accuracy: 100  pp: 10    machine 
overheat(s)          fire special        power: 130  accuracy: 90   pp: 5     machine 
poison-jab           poison physical     power: 80   accuracy: 100  pp: 20    machine 
protect              normal status       power: N/A  accuracy: N/A  pp: 10    machine 
rest                 psychic status      power: N/A  accuracy: N/A  pp: 5     machine 
roar                 normal status       power: N/A  accuracy: N/A  pp: 20    machine 
rock-slide           rock physical       power: 75   accuracy: 90   pp: 10    machine 
rock-tomb            rock physical       power: 60   accuracy: 95   pp: 15    machine 
scorching-sands      ground special      power: 70   accuracy: 100  pp: 10    machine 
shadow-claw          ghost physical      power: 70   accuracy: 100  pp: 15    machine 
sleep-talk           normal status       power: N/A  accuracy: N/A  pp: 10    machine 
solar-beam           grass special       power: 120  accuracy: 100  pp: 10    machine 
stone-edge           rock physical       power: 100  accuracy: 80   pp: 5     machine 
substitute           normal status       power: N/A  accuracy: N/A  pp: 10    machine 
sunny-day(s)         fire status         power: N/A  accuracy: N/A  pp: 5     machine 
swift                normal special      power: 60   accuracy: N/A  pp: 20    machine 
swords-dance         normal status       power: N/A  accuracy: N/A  pp: 20    machine 
take-down            normal physical     power: 90   accuracy: 85   pp: 20    machine 
temper-flare(s)      fire physical       power: 75   accuracy: 100  pp: 10    machine 
tera-blast           normal special      power: 80   accuracy: 100  pp: 10    machine 
thief                dark physical       power: 60   accuracy: 100  pp: 25    machine 
thunder-punch        electric physical   power: 75   accuracy: 100  pp: 15    machine 
u-turn               bug physical        power: 70   accuracy: 100  pp: 20    machine 
upper-hand(s)        fighting physical   power: 65   accuracy: 100  pp: 15    machine 
uproar               normal special      power: 90   accuracy: 100  pp: 10    machine 
vacuum-wave(s)       fighting special    power: 40   accuracy: 100  pp: 30    machine 
will-o-wisp(s)       fire status         power: N/A  accuracy: 85   pp: 15    machine
